### PR TITLE
Add pagination to resources page

### PR DESCRIPTION
**Pull request by cuahsimarko**
### Pull Request Checklist: 
- [ ] Positive Test Case Written by Dev

- [ ] Automated Testing

- [ ] Sufficient User and Developer Documentation

- [ ] Passing Jenkins Build

- [ ] Peer Code review and approval

### Positive Test Case
1. [Enter positive test case here]

### Description
First pass, still need to clean up JavaScript logic for calculating "Filter" counts. Also, requesting of the team that I add pytest and then write unit/functional tests in a separate PR. It's been enough years since I coded JS/frontend that I lost count, so if I did something silly, let me know!

### Related Issues
Fixes https://github.com/hydroshare/hydroshare/issues/2870

### CODE REVIEW

**Comment by cuahsimarko** [comment]
screen shot 2019-02-13 at 3 18 51 pm

**Comment by martinseul** [comment]
thanks Mark, one thing I noticed is that the formatting is different form the formatting used for the pagination on the discover page 

capture2


we should be consistent.

**Comment by cuahsimarko** [comment]
> thanks Mark, one thing I noticed is that the formatting is different form the formatting used for the pagination on the discover page
> 
> capture2
> 
> we should be consistent.

Would you be open to me updating Discover to look like this. I think this is more readable/better UX.

File: hs_core/views/__init__.py [hunk 1]
```diff
@@ -1155,9 +1156,28 @@ def update(self, group_to_update, request):
 @processor_for('my-resources')
 @login_required
 def my_resources(request, page):
-
+    """
+    TODO parameterize and give user control of listings per page, saved in user preferences
```
Comment (pkdash) [inline hs_core/views/__init__.py]: Like it.

File: hs_core/views/__init__.py [hunk 2]
```diff
@@ -1155,9 +1156,28 @@ def update(self, group_to_update, request):
...
+    Template processor for my-resources.html resources listing for user
+    :param request: WSGI GET request for endpoint
+    :param page: Mezzanine middleware in process_view sends <Page: My Resources>
+    :return: html template render with page of resources
+    """
     resource_collection = get_my_resources_list(request)
-    context = {'collection': resource_collection}
+    page = request.GET.get('page')
+    # TODO investigate iRODS performance issue; listing caps out at 10 resources per page, otherwise very slow loading
```
Comment (pkdash) [inline hs_core/views/__init__.py]: I am not sure how it is related to iRODS. If there is any performance issue it has to be database query related.

File: hs_core/views/__init__.py [hunk 3]
```diff
@@ -1155,9 +1156,28 @@ def update(self, group_to_update, request):
...
+    paginator = Paginator(resource_collection, 10)
```
Comment (pkdash) [inline hs_core/views/__init__.py]: Let me know if I am wrong. The benefit we are getting using the pagination is that the user doesn't have to scroll through a long list and the page loading time is quicker as the template engine now has to process only 10 records.

File: hs_core/templates/pages/my-resources.html [hunk 4]
```diff
@@ -24,12 +24,12 @@ <h4 class="panel-title">
                 <div id="filter" class="facet-list panel-collapse collapse in">
```
Comment (pkdash) [inline hs_core/templates/pages/my-resources.html]: I would prefer @Maurier to review the UI code changes.

File: hs_core/views/__init__.py (hunk 3 above)
Comment (cuahsimarko) [reply hs_core/views/__init__.py]: It became a higher priority, due to 30+ second load times for CZO data, due to high number of complex resources, and your statement is correct

File: hs_core/views/__init__.py (hunk 2 above)
Comment (cuahsimarko) [reply hs_core/views/__init__.py]: Understood. I conflated this issue with other iRODS performance issues I am facing when bulk creating/reading entries for CZO activities.

**Comment by dtarb** [comment]
This is a front end design change that has not had any issue written that describes the intent, nor any discussion with the team.  Scrolling down is generally much easier than pagination.  I think this needs broader presentation to the team before merging, e.g. presentation of a quick demo on a call.  If this really solves the load time problem it may be a necessary evil.

**Comment by cuahsimarko** [comment]
David, so this is linked to an Issue, however I didn't see the old comments until just now, mentioning your name. I would have reached out directly, if I had seen that. Yes, this is the only way to get a presentable display to CZOs, as they are evaluating data migration progress. Also, I've been approached by many people, asking for pagination and I lost track of who specifically at USU and or CUAHSI I had spoken to about this issue. Let's get on a call to discuss. I don't think this will make the 1.19 release, anyway, so may just need to be a branch release to CZOs for data preview.

**Comment by cuahsimarko** [comment]
David - Mauriel, Drew, Pabitra, Liza, Tony, Neal, Martin, Scott have all been involved in this work. Mauriel, Pabitra and Scott were involved from the very beginning. Also, this being my first foray into the frontend for HydroShare, I'm still learning the process, which is fluid in my understanding.

**Comment by horsburgh** [comment]
I don't love pagination in general, but the current load time on the My Resources page for me is about 25 seconds and that's turning out to be a bit of a deal killer. I interact with that page all the time, so obviously this gets tedious quickly. Also this has been around for a long time - see #2368 where we recorded some potential options for speeding up the load and where @Maurier posted some profiling information.

File: hs_core/templates/pages/my-resources.html [hunk 5]
```diff
@@ -24,12 +24,12 @@ <h4 class="panel-title">
...
                     <div class="panel-body">
                         <div class="list-group">
-                            <ul class="list-group inputs-group" data-owned-count="{{ owned|length }}"
-                            data-shared-count="{{ shared|length }}" data-added-count="{{ discovered|length }}" >
+                            <ul class="list-group inputs-group"
+                                data-shared-count="{{ shared|length }}" data-added-count="{{ discovered|length }}">
                                 <li class="list-group-item">
-                                    <span data-facet="owned" class="badge">0</span>
+                                    <span class="badge">{{ numOwned }}</span>
```
Comment (Maurier) [inline hs_core/templates/pages/my-resources.html]: I like this. Ideally we should populate the other ones with context variables from the back end too.

**Comment by pkdash** [comment]
We probably are not going to see much improvement in page load time with this pagination as the pagination is mostly at the template level and not at the database level.  Later in a second round of improvement to page loading time we need to look more into the backend code.

**Comment by cuahsimarko** [comment]
We have more extensive test data, so the difference is more obvious. Load times go from 30 seconds to 3-4 seconds.

**Comment by pkdash** [comment]
@cuahsimarko 

> We have more extensive test data, so the difference is more obvious. Load times go from 30 seconds to 3-4 seconds.

In that case it is pretty good. I was not expecting that much improvement during template rendering.

**Comment by horsburgh** [comment]
@cuahsimarko - that improvement would make me really happy. Would it be possible to add a control so the user could specify the number of resources shown in each page? Defaulting to 10 is fine, but I find when I'm using some websites (ha ha, mostly online shopping) I like to view more results per page. I like it when I have control over this. Perhaps something to think about for future improvements.

**Comment by cuahsimarko** [comment]
I have a unique setup, with the new developer setup, iRODS configuration and CZO bulk migration script and data. I was able to test before even pushing the branch to origin. I wasn't optimistic, because I was thinking the same thing you were, but when it worked I was just relieved, since the CZOs are already using those machines as a part of orientation.

**Comment by cuahsimarko** [comment]
> @cuahsimarko - that improvement would make me really happy. Would it be possible to add a control so the user could specify the number of resources shown in each page? Defaulting to 10 is fine, but I find when I'm using some websites (ha ha, mostly online shopping) I like to view more results per page. I like it when I have control over this. Perhaps something to think about for future improvements.

Sure can. I'll do something with local_settings.py for now and shoot you an e-mail how to use it.

**Comment by cuahsimarko** [comment]
@ndebuhr This isn't going out just yet but if you monitor this PR, you can coordinate your Selenium scripts refactor.

**Comment by dtarb** [comment]
@cuahsimarko a user does not have control over local_settings.py.  If we are using paging, a user needs to be able to set the number on the interface (and for me a "show all" option would be great.

**Comment by cuahsimarko** [comment]
@dtarb I'm referring to local developer setups, until the #TODO items are addressed in the code. This was addressed by the team as a part of Pull Request review https://github.com/hydroshare/hydroshare/pull/3148/files

**Comment by cuahsimarko** [comment]
@pkdash @Maurier Can someone re-approve this PR after the small formatting fixes in 761b5002513530631a3377bfcb4ced83246643db and 1fc147e5a168845f4c55ba37bad5cc8e106ecb6e

I'll leave it to you to coordinate with Jeff and Dave to devs to iterate on this and add additional functionality... in separate issues/PRs. This is in the Spirit of Agile Software Development. Until the slowness issue can be resolved, I am suspicious of the merit of selecting higher per page values.

**Comment by cuahsimarko** [comment]
@martinseul for your visibility, calling your attention to this PR and "Agile" and iterating for modern software dev practices, especially where such extensive tech debt and negative User Experience is involved.

**Comment by pkdash** [comment]
@cuahsimarko I agree that we need to put "Agile' into practice.

**Pull request merged**